certbot/error_handler.py

```diff
@@ -34,29 +42,40 @@ class works best as a context manager. For example:
     if a signal is encountered, cleanup_func is called followed by the
     previously registered signal handler.
 
-    Every registered function is attempted to be run to completion
-    exactly once. If a registered function raises an exception, it is
-    logged and the next function is called. If a (different) handled
-    signal occurs while calling a registered function, it is attempted
-    to be called again by the next signal handler.
+    Each registered cleanup function is called exactly once. If a registered
+    function raises an exception, it is logged and the next function is called.
+    Signals received while the registered functions are executing are
+    deferred until they finish.
 
     """
     def __init__(self, func=None, *args, **kwargs):
+        self.body_executed = False
         self.funcs = []
         self.prev_handlers = {}
+        self.received_signals = []
         if func is not None:
             self.register(func, *args, **kwargs)
 
     def __enter__(self):
         self.set_signal_handlers()
 
     def __exit__(self, exec_type, exec_value, trace):
+        self.body_executed = True
```

Comment:
Signals that occur if `__exit__` is triggered due to an exception other than `SignalExit` are completely ignored. This should be fixed.

A nice way to do this I think is to refactor this function a bit to have a form like the following:
```
self.body_executed = True
retval = False
if exec_type in (None, SystemExit):
    return retval
elif exec_type is SignalExit:
    log signals
    retVal = True
else:
    log exception

call_registered
call_signals
reset_signals
return retVal
```